<a href="https://colab.research.google.com/github/ShreyasJothish/DS-Unit-2-Sprint-3-Advanced-Regression/blob/master/DS_Unit_2_Sprint_Challenge_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Science Unit 2 Sprint Challenge 3

## Logistic Regression and Beyond

In this sprint challenge you will fit a logistic regression modeling the probability of an adult having an income above 50K. The dataset is available at UCI:

https://archive.ics.uci.edu/ml/datasets/adult

Your goal is to:

1. Load, validate, and clean/prepare the data.
2. Fit a logistic regression model
3. Answer questions based on the results (as well as a few extra questions about the other modules)

Don't let the perfect be the enemy of the good! Manage your time, and make sure to get to all parts. If you get stuck wrestling with the data, simplify it (if necessary, drop features or rows) so you're able to move on. If you have time at the end, you can go back and try to fix/improve.

### Hints

It has a variety of features - some are continuous, but many are categorical. You may find [pandas.get_dummies](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.get_dummies.html) (a method to one-hot encode) helpful!

The features have dramatically different ranges. You may find [sklearn.preprocessing.minmax_scale](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.minmax_scale.html#sklearn.preprocessing.minmax_scale) helpful!

## Part 1 - Load, validate, and prepare data

The data is available at: https://archive.ics.uci.edu/ml/datasets/adult

Load it, name the columns, and make sure that you've loaded the data successfully. Note that missing values for categorical variables can essentially be considered another category ("unknown"), and may not need to be dropped.

You should also prepare the data for logistic regression - one-hot encode categorical features as appropriate.

In [0]:
# Generic imports
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [0]:
# Load the data from the url considering ' ?' as NaN values
income_url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data'
columns = ['age', 
'workclass',
'fnlwgt',
'education',
'education-num',
'marital-status',
'occupation',
'relationship',
'race',
'sex',
'capital-gain',
'capital-loss',
'hours-per-week',
'native-country',
'income']

income_df = pd.read_csv(income_url, header=None, names=columns, na_values=' ?')

In [3]:
# Check the columns with of NaN values
income_df.isnull().sum()

age                  0
workclass         1836
fnlwgt               0
education            0
education-num        0
marital-status       0
occupation        1843
relationship         0
race                 0
sex                  0
capital-gain         0
capital-loss         0
hours-per-week       0
native-country     583
income               0
dtype: int64

In [4]:
# Update the income column to represent 0 or 1 instead of >50K, <=50K
print(income_df.income.value_counts())
income_df['income'] = income_df['income'].replace({' <=50K': 0, ' >50K': 1})
print(income_df.income.value_counts())

 <=50K    24720
 >50K      7841
Name: income, dtype: int64
0    24720
1     7841
Name: income, dtype: int64


In [5]:
income_df.dtypes

age                int64
workclass         object
fnlwgt             int64
education         object
education-num      int64
marital-status    object
occupation        object
relationship      object
race              object
sex               object
capital-gain       int64
capital-loss       int64
hours-per-week     int64
native-country    object
income             int64
dtype: object

In [0]:
pd.set_option('display.max_columns', None)  # Unlimited columns
pd.set_option('display.max_rows', None)  # Unlimited rows

In [7]:
# One hot encoding of categorical data using get_dummies
obj_columns = ['workclass',
'education',
'marital-status',
'occupation',
'relationship',
'race',
'sex',
'native-country']

income_dummy_df = pd.get_dummies(income_df[obj_columns], dummy_na=True)
print(income_dummy_df.columns)
print(len(income_dummy_df.columns))
income_df.drop(columns=obj_columns, inplace=True)
print(income_df.columns)
print(len(income_df.columns))

Index(['workclass_ Federal-gov', 'workclass_ Local-gov',
       'workclass_ Never-worked', 'workclass_ Private',
       'workclass_ Self-emp-inc', 'workclass_ Self-emp-not-inc',
       'workclass_ State-gov', 'workclass_ Without-pay', 'workclass_nan',
       'education_ 10th',
       ...
       'native-country_ Puerto-Rico', 'native-country_ Scotland',
       'native-country_ South', 'native-country_ Taiwan',
       'native-country_ Thailand', 'native-country_ Trinadad&Tobago',
       'native-country_ United-States', 'native-country_ Vietnam',
       'native-country_ Yugoslavia', 'native-country_nan'],
      dtype='object', length=107)
107
Index(['age', 'fnlwgt', 'education-num', 'capital-gain', 'capital-loss',
       'hours-per-week', 'income'],
      dtype='object')
7


In [8]:
# Join the dataframe to include one hot encoded categorical data
income_cleaned_df = income_df.join(income_dummy_df)
print(len(income_cleaned_df.columns))
income_cleaned_df.describe()

114


age        fnlwgt  education-num  capital-gain  capital-loss  \
count  32561.000000  3.256100e+04   32561.000000  32561.000000  32561.000000   
mean      38.581647  1.897784e+05      10.080679   1077.648844     87.303830   
std       13.640433  1.055500e+05       2.572720   7385.292085    402.960219   
min       17.000000  1.228500e+04       1.000000      0.000000      0.000000   
25%       28.000000  1.178270e+05       9.000000      0.000000      0.000000   
50%       37.000000  1.783560e+05      10.000000      0.000000      0.000000   
75%       48.000000  2.370510e+05      12.000000      0.000000      0.000000   
max       90.000000  1.484705e+06      16.000000  99999.000000   4356.000000   

       hours-per-week        income  workclass_ Federal-gov  \
count    32561.000000  32561.000000            32561.000000   
mean        40.437456      0.240810                0.029483   
std         12.347429      0.427581                0.169159   
min          1.000000      0.000000                0.000000   
25%         40.000000      0.000000                0.000000   
50%         40.000000      0.000000                0.000000   
75%         45.000000      0.000000                0.000000   
max         99.000000      1.000000                1.000000   

       workclass_ Local-gov  workclass_ Never-worked  workclass_ Private  \
count          32561.000000             32561.000000        32561.000000   
mean               0.064279                 0.000215            0.697030   
std                0.245254                 0.014661            0.459549   
min                0.000000                 0.000000            0.000000   
25%                0.000000                 0.000000            0.000000   
50%                0.000000                 0.000000            1.000000   
75%                0.000000                 0.000000            1.000000   
max                1.000000                 1.000000            1.000000   

       workclass_ Self-emp-inc  workclass_ Self-emp-not-inc  \
count             32561.000000                 32561.000000   
mean                  0.034274                     0.078038   
std                   0.181935                     0.268236   
min                   0.000000                     0.000000   
25%                   0.000000                     0.000000   
50%                   0.000000                     0.000000   
75%                   0.000000                     0.000000   
max                   1.000000                     1.000000   

       workclass_ State-gov  workclass_ Without-pay  workclass_nan  \
count          32561.000000            32561.000000   32561.000000   
mean               0.039864                0.000430       0.056386   
std                0.195642                0.020731       0.230670   
min                0.000000                0.000000       0.000000   
25%                0.000000                0.000000       0.000000   
50%                0.000000                0.000000       0.000000   
75%                0.000000                0.000000       0.000000   
max                1.000000                1.000000       1.000000   

       education_ 10th  education_ 11th  education_ 12th  education_ 1st-4th  \
count     32561.000000     32561.000000     32561.000000        32561.000000   
mean          0.028654         0.036086         0.013298            0.005160   
std           0.166834         0.186507         0.114550            0.071646   
min           0.000000         0.000000         0.000000            0.000000   
25%           0.000000         0.000000         0.000000            0.000000   
50%           0.000000         0.000000         0.000000            0.000000   
75%           0.000000         0.000000         0.000000            0.000000   
max           1.000000         1.000000         1.000000            1.000000   

       education_ 5th-6th  education_ 7th-8th  education_ 9th  \
count        32561.000000        32561.000000    32561.00

## Part 2 - Fit and present a Logistic Regression

Your data should now be in a state to fit a logistic regression. Use scikit-learn, define your `X` (independent variable) and `y`, and fit a model.

Then, present results - display coefficients in as interpretible a way as you can (hint - scaling the numeric features will help, as it will at least make coefficients more comparable to each other). If you find it helpful for interpretation, you can also generate predictions for cases (like our 5 year old rich kid on the Titanic) or make visualizations - but the goal is your exploration to be able to answer the question, not any particular plot (i.e. don't worry about polishing it).

It is *optional* to use `train_test_split` or validate your model more generally - that is not the core focus for this week. So, it is suggested you focus on fitting a model first, and if you have time at the end you can do further validation.

In [0]:
# Preprocessing and Regression imports
from sklearn.preprocessing import minmax_scale
from sklearn.linear_model import LogisticRegression

In [10]:
# Spliting the data into independent and dependent variables.
X = income_cleaned_df.drop(columns='income')
y = income_cleaned_df['income']


X_columns = X.columns
X.head(5)

age  fnlwgt  education-num  capital-gain  capital-loss  hours-per-week  \
0   39   77516             13          2174             0              40   
1   50   83311             13             0             0              13   
2   38  215646              9             0             0              40   
3   53  234721              7             0             0              40   
4   28  338409             13             0             0              40   

   workclass_ Federal-gov  workclass_ Local-gov  workclass_ Never-worked  \
0                       0                     0                        0   
1                       0                     0                        0   
2                       0                     0                        0   
3                       0                     0                        0   
4                       0                     0                        0   

   workclass_ Private  workclass_ Self-emp-inc  workclass_ Self-emp-not-inc  \
0                   0                        0                            0   
1                   0                        0                            1   
2                   1                        0                            0   
3                   1                        0                            0   
4                   1                        0                            0   

   workclass_ State-gov  workclass_ Without-pay  workclass_nan  \
0                     1                       0              0   
1                     0                       0              0   
2                     0                       0              0   
3                     0                       0              0   
4                     0                       0              0   

   education_ 10th  education_ 11th  education_ 12th  education_ 1st-4th  \
0                0                0                0                   0   
1                0                0                0                   0   
2                0                0                0                   0   
3                0                1                0                   0   
4                0                0                0                   0   

   education_ 5th-6th  education_ 7th-8th  education_ 9th  \
0                   0                   0               0   
1                   0                   0               0   
2                   0                   0               0   
3                   0                   0               0   
4                   0                   0               0   

   education_ Assoc-acdm  education_ Assoc-voc  education_ Bachelors  \
0                      0                     0                     1   
1                      0                     0                     1   
2                      0                     0                     0   
3                      0                     0                     0   
4                      0                     0                     1   

   education_ Doctorate  education_ HS-grad  education_ Masters  \
0                     0                   0                   0   
1                     0                   0                   0   
2                     0                   1                   0   
3                     0                   0                   0   
4                     0                   0                   0   

   education_ Preschool  education_ Prof-school  education_ Some-college  \
0                     0                       0                        0   
1                     0                       0                        0   
2                     0                       0                        0   
3                     0                       0                        0   
4                     0                       0                        0   

   education_nan  marital-status_ Divorced  marital-status_ Married-AF-spouse  \
0            

In [11]:
# Scaling the Dataframe for easier interpretation of co-efficients.
X = pd.DataFrame(minmax_scale(X), columns=X_columns)
X.head(5)

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: DataConversionWarning: Data with input dtype uint8, int64 were all converted to float64.
  """Entry point for launching an IPython kernel.


age    fnlwgt  education-num  capital-gain  capital-loss  \
0  0.301370  0.044302       0.800000       0.02174           0.0   
1  0.452055  0.048238       0.800000       0.00000           0.0   
2  0.287671  0.138113       0.533333       0.00000           0.0   
3  0.493151  0.151068       0.400000       0.00000           0.0   
4  0.150685  0.221488       0.800000       0.00000           0.0   

   hours-per-week  workclass_ Federal-gov  workclass_ Local-gov  \
0        0.397959                     0.0                   0.0   
1        0.122449                     0.0                   0.0   
2        0.397959                     0.0                   0.0   
3        0.397959                     0.0                   0.0   
4        0.397959                     0.0                   0.0   

   workclass_ Never-worked  workclass_ Private  workclass_ Self-emp-inc  \
0                      0.0                 0.0                      0.0   
1                      0.0                 0.0                      0.0   
2                      0.0                 1.0                      0.0   
3                      0.0                 1.0                      0.0   
4                      0.0                 1.0                      0.0   

   workclass_ Self-emp-not-inc  workclass_ State-gov  workclass_ Without-pay  \
0                          0.0                   1.0                     0.0   
1                          1.0                   0.0                     0.0   
2                          0.0                   0.0                     0.0   
3                          0.0                   0.0                     0.0   
4                          0.0                   0.0                     0.0   

   workclass_nan  education_ 10th  education_ 11th  education_ 12th  \
0            0.0              0.0              0.0              0.0   
1            0.0              0.0              0.0              0.0   
2            0.0              0.0              0.0              0.0   
3            0.0              0.0              1.0              0.0   
4            0.0              0.0              0.0              0.0   

   education_ 1st-4th  education_ 5th-6th  education_ 7th-8th  education_ 9th  \
0                 0.0                 0.0                 0.0             0.0   
1                 0.0                 0.0                 0.0             0.0   
2                 0.0                 0.0                 0.0             0.0   
3                 0.0                 0.0                 0.0             0.0   
4                 0.0                 0.0                 0.0             0.0   

   education_ Assoc-acdm  education_ Assoc-voc  education_ Bachelors  \
0                    0.0                   0.0                   1.0   
1                    0.0                   0.0                   1.0   
2                    0.0                   0.0                   0.0   
3                    0.0                   0.0                   0.0   
4                    0.0                   0.0                   1.0   

   education_ Doctorate  education_ HS-grad  education_ Masters  \
0                   0.0                 0.0                 0.0   
1                   0.0                 0.0                 0.0   
2                   0.0                 1.0                 0.0   
3                   0.0                 0.0                 0.0   
4                   0.0                 0.0                 0.0   

   education_ Preschool  education_ Prof-school  education_ Some-college  \
0                   0.0                     0.0                      0.0   
1                   0.0                     0.0                      0.0   
2                   0.0                     0.0                      0.0   
3                   0.0                     0.0                      0.0   
4                   0.0                     0.0                      0.0   

   education_nan  marital-status_ Divorced  marital-statu

In [12]:
# Fitting the Logistic Regression Model without splitting into test and train set
log_reg = LogisticRegression().fit(X, y)
log_reg.score(X, y)

/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


0.8520929946868954

In [13]:
# Matching coeffients with columns
print('Feature' + '\t' +  'Feature.Coef_')
for i in range(len(X.columns)):
  print(str(X.columns[i]) + '\t' +  str(log_reg.coef_[0][i]))

Feature	Feature.Coef_
age	1.8207854452982375
fnlwgt	0.9565741871549875
education-num	1.7282900565172292
capital-gain	18.020303604207612
capital-loss	2.5613551134693875
hours-per-week	2.772691785833835
workclass_ Federal-gov	0.33108009117801324
workclass_ Local-gov	-0.3222622423250537
workclass_ Never-worked	-0.17777589148321143
workclass_ Private	-0.14989786684600262
workclass_ Self-emp-inc	0.03458997950430413
workclass_ Self-emp-not-inc	-0.6208038903769671
workclass_ State-gov	-0.461279981379163
workclass_ Without-pay	-1.1148775133104738
workclass_nan	-0.3313444184377732
education_ 10th	-0.6639069491392109
education_ 11th	-0.6658414235131123
education_ 12th	-0.3845395219514905
education_ 1st-4th	-0.6206631582293318
education_ 5th-6th	-0.511824828995904
education_ 7th-8th	-0.8713280915269385
education_ 9th	-0.7556598500288404
education_ Assoc-acdm	-0.04248735382495114
education_ Assoc-voc	0.09622014222025796
education_ Bachelors	0.4319019991225167
education_ Doctorate	1.082921653166429

In [0]:
# Additional checking

from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score

# Split the data into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

In [15]:
# Using Logistic Regression to fit the model
model = LogisticRegression()
model.fit(X_train, y_train)

/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='warn',
          n_jobs=None, penalty='l2', random_state=None, solver='warn',
          tol=0.0001, verbose=0, warm_start=False)

In [0]:
# Predict the dependent variable 
y_test_predict = model.predict(X_test)

In [17]:
# Calculate the Mean Squared Error and R2 by comparing prediction with actual values.
MSE = mean_squared_error(y_test, y_test_predict)
RMSE = (np.sqrt(MSE))
print('RMSE is {}'.format(RMSE))

R2 = r2_score(y_test, y_test_predict)
print('R^2 is {}'.format(R2))

RMSE is 0.3782893243566958
R^2 is 0.20795343190268434


## Part 3 - Analysis, Interpretation, and Questions

### Based on your above model, answer the following questions

1. What are 3 features positively correlated with income above 50k?

**Answer:** Based on the sorted coef output of Logistic Regression the top 3 features which are positively corelated with income above 50 K are

|Feature|Coefficient output of Logistic Regression|
|------------|------------------------------------------------------------------|
|capital-gain|18.0203036|
|hours-per-week|2.772691786|
|capital-loss|2.561355113|

2. What are 3 features negatively correlated with income above 50k?

**Answer:** Based on the sorted coef output of Logistic Regression the top 3 features which are negatively corelated with income above 50 K are

|Feature|Coefficient output of Logistic Regression|
|------------|------------------------------------------------------------------|
|sex_ Female|-1.818517409|
|occupation_ Priv-house-serv|	-1.580476773|
|marital-status_ Never-married|	-1.388197543|

3. Overall, how well does the model explain the data and what insights do you derive from it?

**Answer:** The score output of LogisticRegression which gived Mean accuracy of prediction of independent variables or fearures wrt. dependent variable  is 0.8520929946868954.

This signifies an overall good accuracy for the given data set.

Some key take away based on interpretation of top coefficient output are:

* Capital Gain and Capital Loss features are high indications of having income above 50 K. This makes a good sense since the amount of Capital gain or loss is direct indicator of income.
* Also higher hours per week means more income.
* Being female does correlate to having lesser income which seems to be norm in industry and Female right groups actively fight against this disparity. Similarly not being married directly indicates single source of income. So both these features have high negative correlation w.r.t income above 50K.
* Private house servent are having lesser pay which can negatively correlate to having lesser income.

*These answers count* - that is, make sure to spend some time on them, connecting to your analysis above. There is no single right answer, but as long as you support your reasoning with evidence you are on the right track.

Note - scikit-learn logistic regression does *not* automatically perform a hypothesis test on coefficients. That is OK - if you scale the data they are more comparable in weight.

### Match the following situation descriptions with the model most appropriate to addressing them

In addition to logistic regression, a number of other approaches were covered this week. Pair them with the situations they are most appropriate for, and briefly explain why.

Situations:
1. You are given data on academic performance of primary school students, and asked to fit a model to help predict "at-risk" students who are likely to receive the bottom tier of grades.

**Answer:** To model this specific case Quantile Regression would be more apt since we are trying to predict "at-risk" students who shall be at the lower score range instead of mean score range.

With Quantile Regression we can identify the students with respect to median or lower percentile like 35 as compared to Linear Regression which is centered around mean.

2. You are studying tech companies and their patterns in releasing new products, and would like to be able to model and predict when a new product is likely to be launched.

**Answer:** To model this specific case either Survival Analysis or Linear Regression might be useful based on what we are trying to predict.

If we are to predict the new products based on the lifetime of already existing products or features, then Survival Analysis would be apt if end point or death event of previous product is not clear. Otherwise Linear Regression can also be good to predict if previous product cycles are clear with both birth and death event.


3. You are working on modeling expected plant size and yield with a laboratory that is able to capture fantastically detailed physical data about plants, but only of a few dozen plants at a time.

**Answer:** To model this specific case Ridge Regression would be more apt since there is a possibilty of overfitting with fewer sample size. Also since the features are captured in detail there is a possibilty of overfitting with fewer sample size.

To avoid the risk Ridge Regression would be better as compared to Linear Regression.

Approaches:
1. Ridge Regression
2. Quantile Regression
3. Survival Analysis

**TODO - your answers!**